File: src/problems/cheatingstudents/solutions/cheatingstudents_2.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>
#include <set>
using namespace std;

class UnionFind{
    public:
    vector<int> id, sz;
    UnionFind(int n){
        id.resize(n);
        sz.resize(n);
        for (size_t i = 0; i < n; i++){
            id[i] = i;
            sz[i] = 1;
        }
    }
    auto find(int p) -> int {
        int root = p;
        while (root != id[root]) root = id[root];
        while (p != root){
            int next = id[p];
            id[p] = root;
            p = next;
        }
        return root;
    }
    auto connected(int p, int q) -> bool {return find(p) == find(q);}
    auto size(int p) -> int {return sz[find(p)];}
    auto Union(int p, int q) -> void {
        int r1 = find(p);
        int r2 = find(q);
        if (r1 == r2) return;
        if (sz[r1] < sz[r2]){
            sz[r2] += sz[r1];
            id[r1] = r2;
        }else{
            sz[r1] += sz[r2];
            id[r2] = r1;
        }
    }
};

struct Edge{
    int from, to, cost;
    Edge(int f, int t, int c){
        from = f;
        to = t;
        cost = c;
    }
};

auto manhattanDistance(int x, int y, int x1, int y1) -> int {
    return abs(x - x1) + abs(y - y1);
}

auto main() -> int {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    int n; cin >> n;
    vector<pair<int,int>> v(n);
    for (size_t i = 0; i < n; i++) {
        int a, b; cin >> a >> b;
        v[i] = {a, b};
    }
    vector<Edge> edges;
    edges.reserve(n*(n-1)/2);
    for (size_t i = 0; i < n; i++) {
        int x = v[i].first, y = v[i].second;
        for (size_t j = i + 1; j < n; j++) {
            int x1 = v[j].first, y1 = v[j].second;
            edges.emplace_back(Edge(i, j, manhattanDistance(x, y, x1, y1)));
        }
    }
    sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b){
        return a.cost < b.cost;
    });

    int time = 0;
    UnionFind uf = UnionFind(n);
    for(auto& edge : edges){
        int p = edge.from, q = edge.to;
        if(uf.connected(p, q)) continue;
        uf.Union(p, q);
        time += edge.cost;
        if(uf.size(0) == n) break;
    }
    cout << time * 2 << '\n';
    return 0;
}

File: src/problems/cheatingstudents/solutions/cheatingstudents_2.py
import time
class UnionFind:
    def __init__(self, n) -> None:
        self.id = [i for i in range(n)]
        self.sz = [1 for _ in range(n)]
    def find(self, p):
        root = p
        while root != self.id[root]: root = self.id[root]
        while p != root: #path compression
            nxt = self.id[p]
            self.id[p] = root
            p = nxt
        return root
    def connected(self, p, q): return self.find(p) == self.find(q)
    def size(self, p): return self.sz[self.find(p)]
    def union(self, p, q):
        r1 = self.find(p)
        r2 = self.find(q)
        if r1 == r2: return
        if self.sz[r1] < self.sz[r2]:
            self.sz[r2] += self.sz[r1]
            self.id[r1] = r2
        else: 
            self.sz[r1] += self.sz[r2]
            self.id[r2] = r1

class Edge:
    def __init__(self, frm, to, cost) -> None:
        self.frm = frm
        self.to = to
        self.cost = cost

m_d = lambda x, y, x1, y1: abs(x - x1) + abs(y - y1)
n = int(input())
l = []
for i in range(n):
    x, y = map(int, input().split())
    l += [(x, y)]
edges = []
for i in range(n):
    x, y = l[i]
    for j in range(i + 1, n):
        x1, y1 = l[j]
        edges += [Edge(i, j, m_d(x, y, x1, y1))]
start_time = time.time()
edges.sort(key=lambda x: x.cost)
elapsed_time = time.time() - start_time
if elapsed_time > 3: raise RuntimeError
time = 0
uf = UnionFind(n)
for edge in edges:
    p, q = edge.frm , edge.to
    if uf.connected(p, q): continue
    uf.union(p, q)
    time += edge.cost
    if uf.size(0) == n: break
print(time * 2)